### ISSUE TITLE
Needs for frontend and backend package managers are different

### ISSUE THREAD

**Issue by simison** [ipmjs/ipmjs#7]
I.e. that's why we now have NPM and Bower in JS world: 

**At Backend:**
- You want to have hierarchical package structure
- You might have globally installed cli versions, but you might want to use locally installed packages instead (e.g. `glob` & `grunt`) to ensure certain versions.

**At frontend:**
- You don't want to ship multiple versions of a same library, apart from rare exceptions (e.g. Angular v1 and Angular v2 in the same app).
- You want flat package structure
- Because of above, versions might need to be negotiated loosely

More?

**Comment by reimertz** [comment]
For front-end, I argue that bower is unnecessary if you combine a npm-like cli with something like webpack. 

Right now, I use webpack combined with Babel to be able to do 'import $ from "query"', meaning that you'd import dependencies directly where they are needed. 

Webpack make sure to not include the same package twice. This is similar to what we do now on the backend in Java-land.

I'd hope to see people migrate from bower to npm-like cli in the future. :) 

Please correct me if I'm wrong.

**Comment by simison** [comment]
> For front-end, I argue that bower is unnecessary if you combine a npm-like cli with something like webpack.
> 
> Right now, I use webpack combined with Babel to be able to do 'import $ from "query"', meaning that you'd import dependencies directly where they are needed.

While I agree with you, I think it's worth studying all the possible use cases and scenarios to make sure everyone is catered optimally. Just throwing "use webpack" isn't gonna be real world scenario solution, however ideal it would be.

> I'd hope to see people migrate from bower to npm-like cli in the future. :)

Totally. :-) I've always been telling people, it'll be NPM which will replace Bower eventually. Been waiting for that to happen for some years now... doesn't seem to be happening tho, and there must be a reason why.

**Comment by reimertz** [comment]
Ha, touché. :)

I did actually just say "just use webpack". I guess I am one of those hipster devs. :)

Thanks for pointing that out. Totally agree.

**Comment by adelarsq** [comment]
> Totally. :-) I've always been telling people, it'll be NPM which will replace Bower eventually. Been waiting for that to happen for some years now... doesn't seem to be happening tho, and there must be a reason why.

@simison [ied](https://github.com/alexanderGugel/ied) will replace NPM :wink:

**Comment by lbennett-stacki** [comment]
Although there is a relationship between the two, we must define where the package management ends and where the build system starts. For example, with just a hint of modularity, the user should be able to place any package anywhere they want within the scope of that particular root package and then they can configure their build system to either package it all up into a single file to host locally or it can send separate resources to a private CDN. It could be that the user can configure specific package types (think of a modular abstraction of NPM's 'dependencies' vs 'devDependencies' ) which can be customized in a number of ways, including the document root which then allows the package management system to keep track of package locations whilst letting the user organise their file structure ready for the build system?

**Comment by KlonD90** [comment]
No. Bower is shit, one package manager for manage all stuff will be good. If we can reduce some entities we must reduce it. It's more about auto deletion unused parts of code but this is another story which implements in webpack like an example.

**Comment by jamen** [comment]
I think having to develop two different functionalities for one project just because they are in different environments is silly.

I agree with the rest, Browserify or Webpack will do, since they are designed for bundling when that is needed...  

I would go to say that modular in the long run is the best (opinionated), bundle them with a bundler if you don't want it to be.

**Comment by scotttrinh** [comment]
I would just like to "me too" on the idea that shipping code to the browser is really a bundler concern, and the future will be rich with bundling solutions as ES2015 modules start to become a standard convention. Not saying ES2015 modules will solve this now, but bundlers like JSPM, Rollup and the already mentioned Webpack/Browserify is a perfectly good solution for the moment.
